// lib-r-math.js/dist/src/lib/exp/expm1.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _general_1 = require("../common/_general");
var exp = Math.exp, expm1 = Math.expm1, log = Math.log, log1p = Math.log1p;
function R_DT_qIv(lower_tail, log_p, p) {
    return log_p ? (lower_tail ? exp(p) : -expm1(p)) : _general_1.R_D_Lval(lower_tail, p);
}
exports.R_DT_qIv = R_DT_qIv;
function R_DT_CIv(lower_tail, log_p, p) {
    return log_p ? (lower_tail ? -expm1(p) : exp(p)) : _general_1.R_D_Cval(lower_tail, p);
}
exports.R_DT_CIv = R_DT_CIv;
function R_D_LExp(log_p, x) {
    return log_p ? R_Log1_Exp(x) : log1p(-x);
}
exports.R_D_LExp = R_D_LExp;
function R_Log1_Exp(x) {
    if (x > -_general_1.M_LN2) {
        return log(-expm1(x));
    }
    return log1p(-exp(x));
}
exports.R_Log1_Exp = R_Log1_Exp;
function R_DT_Clog(lower_tail, log_p, p) {
    return lower_tail
        ? R_D_LExp(log_p, p)
        : _general_1.R_D_log(log_p, p);
}
exports.R_DT_Clog = R_DT_Clog;
function R_DT_log(lower_tail, log_p, p) {
    return lower_tail ? _general_1.R_D_log(log_p, p) : R_D_LExp(log_p, p);
}
exports.R_DT_log = R_DT_log;
