// lib-r-math.js/dist/src/lib/index.js
'use strict';
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var beta_1 = require("./beta");
exports.Beta = beta_1.Beta;
var binomial_1 = require("./binomial");
exports.Binomial = binomial_1.Binomial;
var binomial_negative_1 = require("./binomial-negative");
exports.NegativeBinomial = binomial_negative_1.NegativeBinomial;
var cauchy_1 = require("./cauchy");
exports.Cauchy = cauchy_1.Cauchy;
var chebyshev_1 = require("./chebyshev");
exports.chebyshev_eval = chebyshev_1.chebyshev_eval;
exports.chebyshev_init = chebyshev_1.chebyshev_init;
var chi_2_1 = require("./chi-2");
exports.ChiSquared = chi_2_1.ChiSquared;
var exp_1 = require("./exp");
exports.Exponential = exp_1.Exponential;
var f_distro_1 = require("./f-distro");
exports.FDist = f_distro_1.FDist;
var gamma_1 = require("./gamma");
exports.Gamma = gamma_1.Gamma;
var geometric_1 = require("./geometric");
exports.Geometric = geometric_1.Geometric;
var hypergeometric_1 = require("./hypergeometric");
exports.HyperGeometric = hypergeometric_1.HyperGeometric;
var logis_1 = require("./logis");
exports.Logistic = logis_1.Logistic;
var lognormal_1 = require("./lognormal");
exports.LogNormal = lognormal_1.LogNormal;
var multinom_1 = require("./multinom");
exports.Multinomial = multinom_1.Multinomial;
var normal_1 = require("./normal");
exports.Normal = normal_1.Normal;
var poisson_1 = require("./poisson");
exports.Poisson = poisson_1.Poisson;
var rng_1 = require("./rng");
exports.rng = rng_1.rng;
exports.IRNG = rng_1.IRNG;
exports.IRNGNormal = rng_1.IRNGNormal;
var signrank_1 = require("./signrank");
exports.SignRank = signrank_1.SignRank;
var student_t_1 = require("./student-t");
exports.StudentT = student_t_1.StudentT;
var tukey_1 = require("./tukey");
exports.Tukey = tukey_1.Tukey;
var uniform_1 = require("./uniform");
exports.Uniform = uniform_1.Uniform;
var weibull_1 = require("./weibull");
exports.Weibull = weibull_1.Weibull;
var wilcoxon_1 = require("./wilcoxon");
exports.Wilcoxon = wilcoxon_1.Wilcoxon;
var log_1 = require("./log");
exports.hypot = log_1.hypot;
exports.log1pR = log_1.log1p;
var bessel_1 = require("./bessel");
var beta_2 = require("./beta");
var common_1 = require("./common");
var gamma_2 = require("./gamma");
exports.special = Object.freeze(__assign({}, gamma_2.default, beta_2.default, common_1.default, bessel_1.default));
var r_func_1 = require("./r-func");
exports.utils = {
    numberPrecision: r_func_1.numberPrecision,
    multiplexer: r_func_1.multiplexer,
    Rcycle: r_func_1.Rcycle,
    flatten: r_func_1.flatten,
    seq_len: r_func_1.seq_len,
    sequenceFactory: r_func_1.sequenceFactory,
    seq1: r_func_1.sequenceFactory(0),
    seq0: r_func_1.sequenceFactory(),
    chain: r_func_1.chain,
    combine: r_func_1.c,
    c: r_func_1.c
};
