// lib-r-math.js/dist/src/lib/rng/index.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var irng_1 = require("./irng");
exports.IRNG = irng_1.IRNG;
var knuth_taocp_1 = require("./knuth-taocp");
var knuth_taocp_2002_1 = require("./knuth-taocp-2002");
var lecuyer_cmrg_1 = require("./lecuyer-cmrg");
var marsaglia_multicarry_1 = require("./marsaglia-multicarry");
var mersenne_twister_1 = require("./mersenne-twister");
var normal_1 = require("./normal");
var inormal_rng_1 = require("./normal/inormal-rng");
exports.IRNGNormal = inormal_rng_1.IRNGNormal;
var super_duper_1 = require("./super-duper");
var timeseed_1 = require("./timeseed");
var wichmann_hill_1 = require("./wichmann-hill");
exports.rng = {
    KnuthTAOCP: knuth_taocp_1.KnuthTAOCP,
    KnuthTAOCP2002: knuth_taocp_2002_1.KnuthTAOCP2002,
    LecuyerCMRG: lecuyer_cmrg_1.LecuyerCMRG,
    MarsagliaMultiCarry: marsaglia_multicarry_1.MarsagliaMultiCarry,
    MersenneTwister: mersenne_twister_1.MersenneTwister,
    normal: {
        AhrensDieter: normal_1.AhrensDieter,
        BoxMuller: normal_1.BoxMuller,
        BuggyKindermanRamage: normal_1.BuggyKindermanRamage,
        Inversion: normal_1.Inversion,
        KindermanRamage: normal_1.KindermanRamage
    },
    SuperDuper: super_duper_1.SuperDuper,
    timeseed: timeseed_1.timeseed,
    WichmannHill: wichmann_hill_1.WichmannHill,
};
