// lib-r-math.js/dist/src/lib/rng/irng.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var IRNG = (function () {
    function IRNG(_seed) {
        this.notify = new Set();
        this.emit = this.emit.bind(this);
        this.register = this.register.bind(this);
        this.unif_rand = this.unif_rand.bind(this);
        this.internal_unif_rand = this.internal_unif_rand.bind(this);
        this.init = this.init.bind(this);
        this._setup();
        this.init(_seed);
    }
    Object.defineProperty(IRNG.prototype, "name", {
        get: function () {
            return this._name;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IRNG.prototype, "kind", {
        get: function () {
            return this._kind;
        },
        enumerable: true,
        configurable: true
    });
    IRNG.prototype.init = function (_seed) {
        this.emit('INIT');
    };
    IRNG.prototype.unif_rand = function (n) {
        var _this = this;
        if (n === void 0) { n = 1; }
        n = (!n || n < 0) ? 1 : n;
        return Array.from({ length: n }).map(function () { return _this.internal_unif_rand(); });
    };
    IRNG.prototype.register = function (event, handler) {
        this.notify.add({ event: event, handler: handler });
    };
    IRNG.prototype.emit = function (event) {
        this.notify.forEach(function (r) {
            if (r.event === event) {
                r.handler();
            }
        });
    };
    return IRNG;
}());
exports.IRNG = IRNG;
